borg/archiver.py

```diff
@@ -405,86 +406,145 @@ def item_is_hardlink_master(item):
+                chunk_id_set1 = set(chunk_ids1)
+                chunk_id_set2 = set(chunk_ids2)
+                total1 = sum_chunk_size(item1)
+                total2 = sum_chunk_size(item2)
+
+                if (can_compare_chunk_ids or total1 != total2 or
+                    not fetch_and_compare_chunks(chunk_ids1, chunk_ids2,
+                                                    archive1, archive2)):
+                    added = sum(c[1] for c in (chunk_id_set2 - chunk_id_set1))
+                    removed = -sum(c[1] for c in (chunk_id_set1 - chunk_id_set2))
+
+                    if total1 is None:
+                        return 'added {:>13}'.format(format_file_size(total2))
+                    elif total2 is None:
+                        return 'removed {:>11}'.format(format_file_size(total1))
+                    else:
+                        return ('{:>9} {:>9}'
+                                .format(format_file_size(added, precision=1, sign=True),
+                                        format_file_size(removed, precision=1, sign=True)))
+
+        def compare_directory(item1, item2):
+            if item2.get(b'deleted') and not item1.get(b'deleted'):
+                return 'removed directory'
+            elif item1.get(b'deleted') and not item2.get(b'deleted'):
+                return 'added directory'
+
+        def compare_owner(item1, item2):
+            user1, group1 = get_owner(item1)
+            user2, group2 = get_owner(item2)
+            if user1 != user2 or group1 != group2:
+                return '[{}:{} -> {}:{}]'.format(user1, group1, user2, group2)
+
+        def compare_mode(item1, item2):
+            if item1[b'mode'] != item2[b'mode']:
+                return '[{} -> {}]'.format(get_mode(item1), get_mode(item2))
+
+        def compare_items(path, item1, item2, hardlink_masters, deleted=False):
             """
             Compare two items with identical paths.
             :param deleted: Whether one of the items has been deleted
             """
+            changes = []
+
+            if item1.get(b'hardlink_master') or item2.get(b'hardlink_master'):
+                hardlink_masters[path] = (item1, item2)
+
+            if has_hardlink_master(item1, hardlink_masters):
+                item1 = hardlink_masters[item1[b'source']][0]
+
+            if has_hardlink_master(item2, hardlink_masters):
+                item2 = hardlink_masters[item2[b'source']][1]
+
+            if get_mode(item1)[0] == 'l' or get_mode(item2)[0] == 'l':
+                changes += filter(None, [compare_link(item1, item2)])
+
+            if b'chunks' in item1 and b'chunks' in item2:
+                changes += filter(None, [compare_content(path, item1, item2,
+                                                         hardlink_masters)])
+
+            if get_mode(item1)[0] == 'd' or get_mode(item2)[0] == 'd':
+                changes += filter(None, [compare_directory(item1, item2)])
```

Comment:
How about `changes.append(compare_XYZ)` and one `filter(None, changes)` in the end?